src/DynamoCore/Core/DynamoMigrator.cs

```diff
@@ -0,0 +1,251 @@
+
+            try
+            {
+                var serializer = new XmlSerializer(typeof(PreferenceSettings));
+
+                using (var fs = new FileStream(preferenceSettingsFilePath, FileMode.Open, FileAccess.Read))
+                {
+                    settings = serializer.Deserialize(fs) as PreferenceSettings;
+                    fs.Close(); // Release file lock
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.Message);
+                Console.WriteLine(ex.StackTrace);
+            }
+            return settings;
+        }
+
+        /// <summary>
+        /// Migrates preference settings and copies packages and custom node 
+        /// definitions from older versions to the currently installed Dynamo version
+        /// </summary>
+        /// <returns></returns>
+        public static PreferenceSettings MigrateBetweenDynamoVersions(string rootFolder)
+        {
+            var assemblyPath = Assembly.GetExecutingAssembly().Location;
+            var currentVersionInfo = FileVersionInfo.GetVersionInfo(assemblyPath);
+
+            // No migration required if the current version is <= version 0.7
+            if (currentVersionInfo.FileMajorPart == 0 &&
+                currentVersionInfo.FileMinorPart <= 7)
+                return null;
+
+            var previousVersion = FindPreviousVersion(rootFolder, currentVersionInfo);
+
+            // No migration required if no previous version exists
+            if (previousVersion.MajorPart == currentVersionInfo.FileMajorPart &&
+                previousVersion.MinorPart == currentVersionInfo.FileMinorPart)
+                return null;
+
+            // Create concrete DynamoMigratorBase object using previousVersion and reflection
+            var sourceMigrator = CreateMigrator(rootFolder, previousVersion.MajorPart, previousVersion.MinorPart);
+            Debug.Assert(sourceMigrator != null);
+            var preferenceSettingsFile = sourceMigrator.ReadPreferences();
+
+            // get migrator object for current version
+            var targetMigrator = CreateMigrator(rootFolder, currentVersionInfo.FileMajorPart, currentVersionInfo.FileMinorPart);
+            Debug.Assert(targetMigrator != null);
+            var preferenceSettings = targetMigrator.WritePreferences(preferenceSettingsFile);
+
+            var sourceDir = Path.Combine(sourceMigrator.CurrentVersionPath, "packages");
+            var targetDir = Path.Combine(targetMigrator.CurrentVersionPath, "packages");
+            Copy(sourceDir, targetDir);
+
+            sourceDir = Path.Combine(sourceMigrator.CurrentVersionPath, "definitions");
+            targetDir = Path.Combine(targetMigrator.CurrentVersionPath, "definitions");
+            Copy(sourceDir, targetDir);
+
+            return preferenceSettings;
+        }
+
+        private static FileVersion FindPreviousVersion(string rootFolder, FileVersionInfo currentVersionInfo)
```

Comment:
Is your concern that the previously installed version may have an empty user data directory, in which case we will have to trigger a chain of migrations? From 0.7 to 0.8, then from 0.8 to 0.9 etc.?